Add Constraint Parser

s9latimm opened the pull request:
Adds a parser for extended constraint syntax.

- [x] implement parser
- [x] add test
- [x] update `FeatureModelParser`

replaces se-sic/vara-feature/pull/28
closes se-passau/VaRA#664

s9latimm committed:
WIP: Implement Parser

A	include/vara/Feature/ConstraintParser.h
M	unittests/Feature/CMakeLists.txt
A	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Fix precedence

M	include/vara/Feature/ConstraintParser.h
M	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Add TODO

M	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Fix tidy

M	include/vara/Feature/ConstraintParser.h
M	unittests/Feature/ConstraintParser.cpp

s9latimm committed:
Ready for review

M	include/vara/Feature/Constraint.h
M	include/vara/Feature/ConstraintParser.h
M	unittests/Feature/ConstraintParser.cpp

vulder committed:
Merge branch 'vara-dev' into f-ConstraintParserNew

s9latimm committed:
Update names

M	include/vara/Feature/ConstraintParser.h

vulder commented on include/vara/Feature/ConstraintParser.h:
```diff
@@ -0,0 +1,492 @@
+  parseConstraint(int NestingLevel = 0,
+                  ConstraintToken::PrecedenceTy Precedence =
+                      ConstraintToken::MaxPrecedence) {
+    if (auto LHS = parseUnaryConstraint(NestingLevel)) {
```
You don't really need the if here. If LHS is a nullptr while(LHS) would skip it anyhow.

vulder reviewed (approved):
Implementation looks quite good :)

s9latimm committed:
Update FMParser and XMLWriter

M	include/vara/Feature/ConstraintParser.h
M	lib/Feature/FeatureModelParser.cpp
M	lib/Feature/FeatureModelWriter.cpp
M	unittests/Feature/ConstraintParser.cpp
M	unittests/resources/test.xml
